fix: fix the type of the argument of `KeyedMutator` for `populateCache`

Key5n opened the pull request:
# The Problem

I will show you the problem with this [code sandbox](https://codesandbox.io/p/sandbox/swr-mutate-with-populatecache-type-error-9h9ynk).
This container uses [code sandbox introduced in swr docs](https://swr.vercel.app/examples/optimistic-ui) with some changes.

I just added TypeScript and changed like below.

```diff
export async function addTodo(todo) {
  await delay();
  todos = [...todos, todo];
-  return todos;
+ return todo;
}

```

In this codesandbox, there is an serious type error:
```
Argument of type 'Promise<ToDo>' is not assignable to parameter of type 'ToDo[] | Promise<ToDo[] | undefined> | MutatorCallback<ToDo[]> | undefined'.
  Type 'Promise<ToDo>' is not assignable to type 'Promise<ToDo[] | undefined>'.
    Type 'ToDo' is missing the following properties from type 'ToDo[]': length, pop, push, concat, and 25 more.
```
in the source code:
```ts
await mutate(addTodo(newTodo.text), {
  optimisticData: [...(data ?? []), newTodo],
  populateCache: (updatedTodo, todos) => [
    ...(todos ?? []),
    updatedTodo,
  ],
  revalidate: false,
});
```

I referred to [this article](https://github.com/vercel/swr/issues/1817).

In this article, the example written in JavaScript uses bound `mutate` to make `addToDo` return just one updated item, rather than multiple updated items. 
However, when used in TypeScript like the codesandbox I suggested, this results in a type error.
For instance, when useSWR fetches an API that returns a `ToDo[]` type, the first argument of `mutate` is expected to be of type `ToDo[]`.
This works fine in JavaScript, but the issue arises only when it's converted to TypeScript.

# What this PR does
This PR fixes a type error problem related to `populateCache` and bound `mutate`.

Now, the first argument of bound `mutate` only accepts a type of data fetched by `useSWR`.

But in order to make use of `populateCache`, the first argument of `mutate` should accept any type of API fetching result. 

So I changed the type of the first argument from `Data` (which is the type of data `useSWR` fetches) into `MutationData` (which is the type of the data returned by the mutator)

In addition, I changed types in `infinite` since it had error when I changed the type of `KeyedMutator` .

Key5n commented on src/_internal/types.ts:
```diff
@@ -428,21 +428,24 @@ export interface ScopedMutator {
+  <Data = any, MutationData = Data>(
     key: Arguments,
-    data?: T | Promise<T> | MutatorCallback<T>,
-    opts?: boolean | MutatorOptions<Data, T>
-  ): Promise<T | undefined>
+    data?: MutationData | Promise<MutationData> | MutatorCallback<MutationData>,
+    opts?: boolean | MutatorOptions<Data, MutationData>
+  ): Promise<MutationData | undefined>
```
This is just refactoring for naming consistency.
Since there is `@typeParam MutationData` comment docs, the generics parameter T should be follow that.

huozhi reviewed (commented):
Thanks for the PR, can we add a test?

Key5n commented:
Sure!
If you want to modify my PR, it is at your mercy.

Key5n commented:
Sorry, I didn't notice that some tests were unpassed.
I will fix it.

Key5n commented:
@huozhi 

A PR I submitted fails some tests, so I've changed almost of whole things and add tests.
Sorry for begging you reviews again.

I changed the type `KeyedMutator` using overload function, making `populateCache` is necessary when `MutationData` (the type of the first argument of bound mutate) is different from the type of `Data`.

This achieves strict type attaching.

Key5n commented on src/infinite/index.ts:
```diff
@@ -238,16 +237,9 @@ export const infinite = (<Data, Error>(useSWRNext: SWRHook) =>
-          | MutatorCallback<Data[]>,
-        opts?: undefined | boolean | SWRInfiniteMutatorOptions<Data[], T>
-      ) {
+      function (data?: any, opts?: any) {
```
How to solve this problem.

1. use any
2. tell the callback function about the type of `Data` and compare it like the following code 

```ts
// this is invalid code, just for sharing my idea
if (typeof Data === typeof MutationData){
    
} else {
  ...
}
```

3. soft type attaching like the following code

```ts
    const mutate = useCallback<SWRInfiniteKeyedMutator<Data[]>>(
      // eslint-disable-next-line func-names
      function <MutationData = Data[]>(
        data?:
          | Data
          | Promise<Data | undefined>
          | MutatorCallback<Data>
          | MutationData
          | Promise<MutationData | undefined>
          | MutatorCallback<MutationData>
          | undefined,
        opts?: boolean | SWRInfiniteMutatorOptions<Data[], MutationData>
      ): Promise<Data[] | MutationData | undefined> {

       // contents

        return arguments.length
          ? swr.mutate<Data[]>(
              data as any,
              {
                ...options,
                revalidate: shouldRevalidate
              } as any
            )
          : swr.mutate()
      },
      // swr.mutate is always the same reference
      // eslint-disable-next-line react-hooks/exhaustive-deps
      [infiniteKey, cache]
    )
```
